core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
-            dnsMultiCast.close();
-        interruptReloadThread();
+        return pending;
+    }
+
+    private void _cleanUpShutdownUDPBroadcast(List<Throwable> errors) {
+        if(udpBroadcastThread!=null) {
+            LOGGER.log(Level.FINE, "Shutting down {0}", udpBroadcastThread.getName());
+            try {
+                udpBroadcastThread.shutdown();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(SEVERE, "Failed to shutdown UDP Broadcast Thread", e);
+                // safe to ignore and continue for this one
+            } catch (Throwable e) {
+                LOGGER.log(SEVERE, "Failed to shutdown UDP Broadcast Thread", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
+
+    private void _cleanUpCloseDNSMulticast(List<Throwable> errors) {
+        if(dnsMultiCast!=null) {
+            LOGGER.log(Level.FINE, "Closing DNS multi-cast service-discovery");
+            try {
+                dnsMultiCast.close();
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(SEVERE, "Failed to close DNS Multicast service", e);
+                // safe to ignore and continue for this one
+            } catch (Throwable e) {
+                LOGGER.log(SEVERE, "Failed to close DNS Multicast service", e);
+                // save for later
+                errors.add(e);
+            }
+        }
+    }
+
+    private void _cleanUpInterruptReloadThread(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Interrupting reload thread");
+        try {
+            interruptReloadThread();
+        } catch (SecurityException e) {
+            LOGGER.log(WARNING, "Not permitted to interrupt reload thread", e);
+            errors.add(e);
+        } catch (OutOfMemoryError e) {
+            // we should just propagate this, no point trying to log
+            throw e;
+        } catch (LinkageError e) {
+            LOGGER.log(SEVERE, "Failed to interrupt reload thread", e);
+            // safe to ignore and continue for this one
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to interrupt reload thread", e);
+            // save for later
+            errors.add(e);
+        }
+    }
 
-        java.util.Timer timer = Trigger.timer;
-        if (timer != null) {
-            timer.cancel();
+    private void _cleanUpShutdownTriggers(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Shutting down triggers");
+        try {
+            java.util.Timer timer = Trigger.timer;
+            if (timer != null) {
+                timer.cancel();
+            }
+            // TODO: how to wait for the completion of the last job?
```

Comment:
you schedule a task on the timer to run immediately and that task cancels the timer.  This guarantees it is the last task to execute.